### PR TITLE
allowing reporting to happen in forked process

### PR DESCRIPTION
**Pull request by Kallin**
simple forking solution to support performance concerns from https://github.com/danmayer/coverband/issues/124

### CODE REVIEW

**Comment by Kallin** [comment]
please let me know if you'd like to see this done differently or if there's anything else I can do to help get this PR accepted.

**Comment by kbaum** [comment]
This is a very interesting idea. My only thought is that we might get the performance we need by just reporting to redis in a background thread instead:

https://github.com/danmayer/coverband/issues/124#issuecomment-425582122

Then we avoid the added branching logic in the code.

File: lib/coverband/collectors/base.rb [hunk 1]
```diff
@@ -76,10 +77,22 @@ def stop_coverage
         raise 'abstract'
       end
 
-      def report_coverage
+      def report_coverage_potentially_forked
```
Comment (kbaum) [inline lib/coverband/collectors/base.rb]: To avoid the extra "abstract" method, can we pull the forking logic up to the base class?

Comment (kbaum) [reply lib/coverband/collectors/base.rb]: Ok. I think i misread this initially. Apologize for that.

**Comment by kbaum** [comment]
RE: Sampling. With the ruby coverage approach to coverage as opposed to trace, coverage is being tracked all of the time.. whether we report for that request or job execution. Sampling percentage then only controls how often we report to redis. Since the coverage data is global available to the process through [Coverage#peek_result](https://ruby-doc.org/stdlib-2.5.0/libdoc/coverage/rdoc/Coverage.html#method-c-peek_result), there is no need to report to redis during the request.

RE: Threads and ruby. It is true that ruby cannot perform concurrent processing across multiple CPU's. That being said, when one thread is waiting on IO, another thread of execution can run.  This is an old post but still holds up:

https://yehudakatz.com/2010/08/14/threads-in-ruby-enough-already/

In the case of reporting to redis, all we are doing is IO so I would think that this would be an excellent case to leverage threads.

**Comment by danmayer** [comment]
seems lots of folks are interested in the same basic idea, I have started work on coverband 3 to pull things out and simplify so I don't think we want many PRs on the older code base before that... I think forking is interesting and I plan to keep a percentage based reporting and not move entirely to the background thread as I agree there are a number of use cases where it will still be an easier solution... The other thing is really how we are reporting now the thousands of redis calls isn't a good use a single store hash and summing across them only at report time is probably more efficient.

**Comment by danmayer** [comment]
I pulled in the pipeline change, but would like to hold off on these kinds of concepts until `coverband 3` as I think we will be able to significantly improve reporting once simplified

**Comment by Kallin** [comment]
Thanks for taking a look, and I look forward to seeing what shows up in Coverband 3.

Re: a couple of your comments:

Forking & Memory usage: I suppose this may depend on the underlying OS, but my understanding is that Linux and modern OSes implement copy-on-write when forking in order to keep forked processes lightweight, https://unix.stackexchange.com/questions/155017/does-fork-immediately-copy-the-entire-process-heap-in-linux . I suppose if we wanted to verify this we could create a bunch of ruby forks that just sit around sleeping for a while and keep forking and see how memory usage increases with each fork.

It would be interesting to compare performance of forking w/ background thread in a 'standard' scenario (I'm thinking rails web request). Forking should add close to no overhead to the time of the request, as the request just forks and never waits. Threading will add some overhead I imagine due to context-switching, but it's hard to know how much. If any of the 'reporting' thread takes a chunk of time other than in low-level IO, the web request would be waiting on it. I wonder what % of the time spent in the reporting is actually low-level IO, and how much is spent in the redis library marshalling/unmarshalling request/response etc., that would not qualify as IO and would block other threads.

It may be a low overhead with the threading, but in our case we have a very strict time limit on certain requests that we want to keep things under, so even adding 100ms or something to a request may be too much. 

Anyways, I'd encourage a deeper look at forking, I'm still bullish on it :)

**Comment by kbaum** [comment]
You make some very compelling points re: forking. One thought is that with the new coverage collector, there is no need to report to redis during a request since coverage data is global for the process. We could even have a background thread that loops, sleeps, forks, and reports.

**Comment by danmayer** [comment]
Yeah I will have to do some tests but my thinking would be for a multi-threaded server you would have the parent process start the thread and do reporting out of request cycle... 

I think the most common target is Puma, which is interesting because of cluster mode...

`Puma also offers "clustered mode". Clustered mode forks workers from a master process. Each child process still has its own thread pool.`

If we can support background reporting in clustered mode across processes and threads I think that should cover all the various use cases... Then we just need to find the various hooks for alternative servers or offer a middleware / forking option for folks on non-standard servers.

**Comment by kbaum** [comment]
I noticed that when unicorn and puma fork, ruby coverage is reset for some reason.
https://github.com/danmayer/coverband/issues/134
Would that apply here?

**Comment by kbaum** [comment]
@Kallin - going to close this for now as this has some conflicts and I am seeing that forking resets coverage in the forked process. We can always reopen once conflicts are resolved if i am wrong on that.

**Pull request closed without merging**